include/evmmax/evmmax.hpp

```diff
@@ -128,5 +128,70 @@ class ModArith
         const auto s = d.value + mod;
         return (d.carry) ? s : d.value;
     }
+
+    /// Compute the modular inversion of the x in Montgomery form. The result is in Montgomery form.
+    /// If x is not invertible, the result is 0.
+    constexpr UintT inv(const UintT& x) const noexcept
+    {
+        assert((mod & 1) == 1);
+        assert(mod >= 3);
+
+        // Precompute inverse of 2 modulo mod: inv2 * 2 % mod == 1.
+        // The 1/2 is inexact division which can be fixed by adding "0" to the numerator
+        // and making it even: (mod + 1) / 2. To avoid potential overflow of (1 + mod)
+        // we rewrite it further to (mod - 1 + 2) / 2 = (mod - 1) / 2 + 1 = ⌊mod / 2⌋ + 1.
+        const auto inv2 = (mod >> 1) + 1;
+
+        // Use extended binary Euclidean algorithm. This evolves variables a and b until a is 0.
+        // Then GCD(x, mod) is in b. If GCD(x, mod) == 1 then the inversion exists and is in v.
+        // This follows the classic algorithm (Algorithm 1) presented in
+        // "Optimized Binary GCD for Modular Inversion".
+        // https://eprint.iacr.org/2020/972.pdf#algorithm.1
+        // TODO: The same paper has additional optimizations that could be applied.
+        UintT a = x;
+        UintT b = mod;
+
+        // Bézout's coefficients are originally initialized to 1 and 0. But because the input x
+        // is in Montgomery form XR the algorithm would compute X⁻¹R⁻¹. To get the expected X⁻¹R,
+        // we need to multiply the result by R². Do this by initializing u to R².
+        UintT u = m_r_squared;
```

Comment:
out of curiosity - not easier to correct for this at the end? Any external source to cite that this variation is equivalent?

There are a lot of modification to the linked classical algorithm, could do away with at least one. The others are kinda easier to warp head around.

Reply:
We borrowed the idea from zksync implementation of some precompiles. Maybe it would be equivalent to do this in the end, but that would be additional multiplication (here this is free). But I will try :)

Reply:
"Fixing" `v` after works, but it requires 2 multiplications by R² (because of the performed reduction):

```cpp
        v = mul(v, m_r_squared);
        v = mul(v, m_r_squared);
        return v;
```

Reply:
Or in other words: `v = mul(v, to_mont(m_r_squared));`.

Reply:
I'm OK with keeping as is, if we have a solid reference to cite and confidence all the edge cases are also correctly handled (and later tested).